### PR TITLE
Rewrite the exhaustiveness specification.

### PR DESCRIPTION
**Pull request by munificent**
This is based on Johnni's excellent implementation. I tried to follow the semantics of the implementation as closely as I could, and mostly stuck to the terminology too. The main change is that the implementation uses "type" to also sort of do some restriction stuff, and I tried to keep that separate here such that a "type" on a space always refers to a static type.

I'm sure there are some mistakes here (but I of course don't know where they are). I imagine it will take some iteration between the implementation, the spec, and the tests before they settle on some kind of fixpoint. But this should hopefully be enough to start writing some co19 tests.

cc @sgrekhov

### CODE REVIEW

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 1]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
-A a = ...
-switch (a) {
-  case B b: ...
-  case F f: ...
-}
-```
+    2.  Else, append a `space` to `result` whose type is `Object?`. *If a
+        pattern doesn't have a subpattern for some field or getter, then it
+        matches all values that the field or getter could return, which is
+        equivalent to it matching it with type `Object?`.*
```
Comment (johnniwinther) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:1079]: While this is true for the case spaces it isn't precise enough for the value space since it might lead to a space larger that what the type of the property can contain. For instance, if the value space is `Object()` and we have a case space `Object(hashCode: int())` then unpacking the key `hashCode` for `Object()` to `Object?()` would mean that we would conclude that `int()` does not exhaust the `hashCode` property. Instead we need to use the type corresponding to the property of the type of the space when it exists. Currently the algorithm uses `Object?()` as a fallback when the property doesn't exist, which (I think) can only happen for erroneous code. (Given that we have checked that the type of the value space is a subtype of the case space, so if the property exists on the case space type it must exist on the value space type.)

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 2]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
-y:    club │(x: ♣︎, y: ♣︎)││(x: ♦︎, y: ♣︎)  (x: ♥︎, y: ♣︎)  (x: ♠︎, y: ♣︎)│
-           │            ││                                        │
-   diamond │(x: ♣︎, y: ♦︎)││(x: ♦︎, y: ♦︎)  (x: ♥︎, y: ♦︎)  (x: ♠︎, y: ♦︎)│
-           │            ││                                        │
-     heart │(x: ♣︎, y: ♥︎)││(x: ♦︎, y: ♥︎)  (x: ♥︎, y: ♥︎)  (x: ♠︎, y: ♥︎)│
-           └────────────┘│                                        │
-     spade    <hole>     │(x: ♦︎, y: ♠︎)  (x: ♥︎, y: ♠︎)  (x: ♠︎, y: ♠︎)│
-                         └────────────────────────────────────────┘
-```
+### Are any switch cases unreachable?
```
Comment (johnniwinther) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:675]: To put this section in context we should probably put at the end of the specification more as a side note and mention that this is an unrequired extension of the exhaustiveness checking that can be used by tools to provide hints.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 3]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+3.  For each `worklist` in `cases`:
 
-## Next steps
+    1.  Dequeue the first case union in `worklist` to `caseUnion`.
 
-I haven't proven that the algorithm is correct. I also haven't proven any bounds
-on its performance. A fairly large test covering every combination of four
-fields seems to be snappy (and certainly was not with earlier versions of this
-algorithm).
+    1.  For each `case` space in `caseUnion`:
```
Comment (johnniwinther) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:1009]: The item number is off here and in the sublist below (not that it matters to the markdown renderer).

**Review by johnniwinther — APPROVED** [review]

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 4]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
 
-```
-FutureOr<int> - int     expands to:  Future<int>|int - int = Future<int>
-FutureOr<int> - Future  expands to:  Future<int>|int - Future = int
-```
+*   **The declaration `D` of `T` is a `sealed` type:**
 
-### Subtraction
+    1.  For each declaration `C` in the library of `D` that has `D` in an
+        `implements`, `extends`, or `with` clause:
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:816]: Here I want to restrict it to:

... declaration `C` in the library of `D` that has `D` as a direct superdeclaration through an `implements`, `extends` or `with` clause _(I agree with excluding `on` types too)_
and which does not have any other superdeclaration *P* which has *D* as a proper superdeclaration.

This avoids weird behavior for unnecessary implements clauses:
```dart
sealed class S {}
class C extend S {}
class D extends S {}
class E extends D implements S {}
```
Here class `E` should *not* be included, because it's a proper subtype of a proper subtype of `S` (if `D` was not `sealed`, it would be included transitiviely in exhausting `S`, but that's not important here).

It's not necessary to include `E` in the set of spaces for `S` here (exhaustiveness of `S` requires exhausting `D` anyway), so it just complicates everything.

(That is: The *longest* path from the subclass to the sealed class in the class declaration hierarchy has length 1. Internal anonymous mixin applications don't count since we use "declaration" here.
Or as you say below, "a direct subtype" - ignoring anonymous mixin types as usual.)

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 5]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+        2.  Else (not a trivial substitution) let `S` be an overapproximation
+            of `C`.
 
-    test(Animal a) {
-      switch (a) {
-        case Mammal m: ...
-      }
-    }
-    ```
+            **TODO:** Is the overapproximation part here correct?
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:832]: What *is* an "overapproximation"?

(Ah, found definition below. Add: `(see definition in "Overapproximation" section below)`)

The overapproximation is defined as a mapping from type to a type, but is probably really from a declaration to a type (types are instantiated, declarations are not), and `S` is a space, not a type. 
Should `S` be the space for that type?

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 6]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+    `Null`.
 
-    1.  Let `Lf` be the field in `L` and `Rf` be the field in `R` (either of
-        which may be inferred if not present).
+*   **`T` is `FutureOr<F>` for some `F`**: A space of type `Future<F>` and a
+    space of type `F`.
 
-    2.  Calculate `D = Lf - Rf`.
+*   **`T` is `bool`:** Two spaces of type `bool` with constant restrictions
+    `true` and `false`.
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:918]: And

*    **`T` is `Null`:** One space of type `bool` with the constant restriction `null`.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 7]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
 
-        ```dart
-        String s = ...
-        switch (s) {
-          case 'a string': ...
-          case 'a string': ...
-        }
-        ```
+*   **Constant pattern:** A space whose type is the type of the constant and
+    with a constant restriction for the given constant value.
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:469]: (Do we require the type of a constant pattern to have primitive equality? If not, should this only apply if it does?)

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 8]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+null
+```
 
-2.  If `L` or `R` is a union, then `I` is empty if the intersection of every
-    arm of the union with the other operand is empty.
+### Overapproximation
 
-3.  Otherwise, we're intersecting two object spaces.
+The *overapproximation* of a type `T` is `T` with all type variables replaced
+with their *defaults*:
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:979]: I think this is slightly underdefined.

I think we only apply "overapproximation" to `class`, `mixin` and `enum` declarations.  So say that.
(It gets weird if we try to replace type variables of a generic function type.)
We can always extend it to apply to the type of such a declaration by going from the type to the declaration.

Also, we should probably say that we apply it to the *declaration*, not the type.
(Types do not have type variables or bounds, their declarations - and, depending on mood, their generic classes and interfaces, which are still different from the *types* - do. The types have type *arguments*, but the variables are long gone.)

So, given a generic enum, mixin or class declaration, instantiate its type with the default types for its declaration's type parameters.
But that's basically what "instantiate to bounds" (I2B) does, giving you a valid super-bounded supertype for all valid instantiations, so maybe just use that?
(I2B also accounts for bound expressions that aren't straight-up types, up-to and including F-bounded type parameters. Can't just use a bound as a type directly.)

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 9]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
 
-    1.  If neither object's type is a subtype of the other then `I` is empty.
-        They are unrelated types with no (known) intersection.
+1.  If `T` is in a contravariant position, then the default is `Never`.
 
-    2.  Otherwise, go through the fields. If the intersection of any
-        corresponding pair of fields is empty, then `I` is empty.
+2.  Otherwise, the default is the bound if there is one.
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:983]: `class C<T, S extends T, R extends List<T>> {}` has bound declarations, but not bound types.
Use I2B, it's good for you.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 10]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
-    arm of the union with the other operand is empty.
+### Overapproximation
 
-3.  Otherwise, we're intersecting two object spaces.
+The *overapproximation* of a type `T` is `T` with all type variables replaced
+with their *defaults*:
 
-    1.  If neither object's type is a subtype of the other then `I` is empty.
-        They are unrelated types with no (known) intersection.
+1.  If `T` is in a contravariant position, then the default is `Never`.
```
Comment (lrhn) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:981]: (Can a type parameter of an `extends`, `with` or `implements` type variable occur contravariantly - at least after alias expansion? Or is this just in preparation for variance annotations? In any case I2B should do the right thing.)

**Review by lrhn — COMMENTED** [review]
Selectively reviewed the spec-like details, not the algorithm.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 8 above)
Comment (johnniwinther) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:979]: I tend to think of this as "recursively instantiate to bounds" and it is slightly different that I2B. I2B works on raw types whereas the overapproximation works on instantiated types.

For instance the overapproximation of `Map<K, Map<V, K Function(K)>>>` in:
```
method<K extends num, V extends Foo<V>>(Map<K, Map<V, K Function(K)>>> map) => switch (map) { ... };
```
is `Map<num, Map<dynamic, num Function(Never)>>>`, so that each type variable occurring in the type has been replaced with the I2B default type modulo variance.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 1 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:1079]: Ah, you're right. I felt like I was getting something wrong here. Fixed. (And added a section up on properties defining how to get the type of a key.)

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 2 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:675]: I can move this if you want, but I think this is the natural place to stitch together how reachability is defined in terms of the core is-exhaustive function.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 11]
```diff
@@ -185,13 +189,14 @@ we would break that principle.
...
 ### Sealing supertypes
 
-Instead, we extend the language to let a user explicitly "seal" a supertype with
-a specified closed set of subtypes. No other code is allowed to define a new
-subtype of the sealed supertype (using `extends`, `implements`, or `with`). The
-supertype is also implicitly made abstract.
+Instead, we [extend the language to let a user explicitly "seal"][sealed] a
+supertype with a specified closed set of subtypes. No other code is allowed to
+define a new subtype of the sealed supertype (using `extends`, `implements`, or
+`with`). The supertype is also implicitly made abstract.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:195]: Done.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 12]
```diff
@@ -185,13 +189,14 @@ we would break that principle.
 
 ### Sealing supertypes
 
-Instead, we extend the language to let a user explicitly "seal" a supertype with
-a specified closed set of subtypes. No other code is allowed to define a new
-subtype of the sealed supertype (using `extends`, `implements`, or `with`). The
-supertype is also implicitly made abstract.
+Instead, we [extend the language to let a user explicitly "seal"][sealed] a
+supertype with a specified closed set of subtypes. No other code is allowed to
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:193]: Done.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 13]
```diff
@@ -185,13 +189,14 @@ we would break that principle.
...
+`with`). The supertype is also implicitly made abstract.
 
-This document does *not* propose a specific syntax for sealing, but we'll use
-this hypothetical syntax for now to show an example:
+[sealed]: https://github.com/dart-lang/language/blob/master/accepted/future-releases/sealed-types/feature-specification.md
+
+Here is an example:
 
 ```dart
 enum Suit { club, diamond, heart, spade }
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:202]: Done.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 14]
```diff
@@ -230,10 +235,10 @@ A parenthesized class name means "sealed". It's important to understand what
...
-    extension or implementation. In the example here, anyone can extend,
-    implement or even mixin, say, `Pip` or `Queen`. This doesn't cause any
-    problems. And in fact, we use this to turn `Face` into a sealed supertype
-    of its own set of subtypes.
+    extension or implementation. In the example here, anyone can extend or
+    implement, say, `Pip` or `Queen`. This doesn't cause any problems. And in
+    fact, we use this to turn `Face` into a sealed supertype of its own set of
+    subtypes.
 
 *   The subtypes do not have to be disjoint. We could define a `Royalty` class
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:243]: Yeah, `Royalty` isn't a good example. Came up with another one.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 15]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
 critical invariant that if we have matched against all instances of those
 subtypes, then we have exhaustively covered all instances of the supertype.
 
-## Types, patterns, and spaces
+## Spaces
 
 The two questions exhaustiveness answers are defined based on some notion of a
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:256]: At the very beginning of the doc, it says:

> Exhaustiveness checking is about answering two questions...

But that's certainly far enough away to be worth clarifying here. Reworded.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 16]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+enough things like the set of values matched by the pattern `DateTime(day: 1)`
+which matches not just values of some type, but also ones whose fields have
+certain values.
 
-### Static types
+A pattern is obviously the natural way to describe a set of values of some type
+filtered by some arbitrary set of predicates on their properties. However, the
+patterns proposal defines a rich set of patterns to make the feature user
+friendly and expressive. It would require unnecessary complexity in the
+exhaustiveness algorithm to handle every single kind of pattern.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:272]: Reworded a bit.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 17]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
 
-In a statically typed language, the first obvious answer is a static type. A
-type does represent a set of values. The type `bool` represents the set `true`
-and `false`. The type `DateTime` represents the set of all instances of the
-`DateTime` class or any of its subtypes.
+Instead, following Liu's paper, we use a data structure called a *space*. Spaces
+are essentially a desugared, simpler representation of a pattern.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:275]: Reworded a bit.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 18]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
-    union pattern to represent "`1` or `2`".
+*   An *open restriction*, the default, matches all values.
 
-2.  On the other hand, patterns are too complex. Because the proposal is
-    intended to be richly expressive, it defines a large number of different
-    kinds of patterns each with its own semantics. Handling every one of those
-    (and every pairwise combination of them) would add a lot of complexity to
-    the algorithm.
+*   A *constant restriction* matches only values that are identical to a given
+    constant value. It's used for enum values, `true`, and `false`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:302]: I used a space whose type is `Null` to handle `null`. Every space has a type anyway, so that's sufficient.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 19]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+*   **Relational pattern:** The empty space union. *Relational patterns don't
+    reliably match any values, so don't help with exhaustiveness.*
+
+*   **Cast pattern:** *A cast pattern matches if the value is of the casted type
+    and the inner pattern matches it. But a cast pattern also "handles" a value
+    when the cast _fails_. From the perspective of exhaustiveness checking, what
+    matters it that execution can't flow out of a switch without matching at
+    least one case. But if a pattern _throws_ a runtime exception, then
+    execution also doesn't flow out of it. So we treat throwing as essentially
+    another kind of matching.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:373]: Done.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 20]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+    test(A a) => switch (a) {
+      C(field: 0) as C => 0,
+      C _ => 1
+    };
     ```
 
-### Types in spaces
+    *After the first case, we know that if `a` is a `B` or `D` then we will have
+    thrown an exception. But if `a` is a `C`, it may not have matched if the
+    field isn't `0`. So the space for the first case is `B|C(field: 0)|D'.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:408]: Good catch. There used to be a lot more in here about notation, but I don't think it's very useful for the new algorithm so I discarded most of it. Add a "union of" to clarify. The notation is informal now and hopefully self evident.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 21]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
-        and the case is unreachable. Otherwise, the case is reachable and
-        `remaining` describes the values it might match.
+2.  Else, if either side is a union, then a space union of the intersection of
+    each operand of the union with the other space.
 
-Note that we can calculate reachability even for cases with guards. It's useful
-to tell if a case with a guard is completely unreachable. It's just that we
-ignore guards on *preceding* cases when determining if each case is reachable.
+    *In other words, distribute the intersection into the branches. So `(A|B) ^
+    C` is `(A^C)|(B^C)`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:571]: Added some prose.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 22]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+3.  Else if both are constant restrictions and the constants are identical then
+    `true`.
 
-```
-Jack(suit: heart|club) - Card(suit: club) = Jack(suit: heart)
-```
+4.  Else if both are arity restrictions:
 
-The objects have different types now, but it still works out. Now consider:
+    1.  If `b.hasRest` then `true` if `a.size >= b.size`. *Since `b` has no
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:621]: I try to be clear about that. In this case, the omission is deliberate. If `b.hasRest` is `false`, it falls through to the next cases.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 23]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+the second switch is clearly exhaustive because every value is a `Face` and the
+second case matches all instances of `Face`.
 
-```
-Face? - Face  expands to:  Face|Null - Face = Null
-Jack? - Null  expands to:  Jack|Null - Null = Jack
-```
+In the algorithm, when it looks at a space, it first *expands* it. If the type
+of a space is a sealed type, enum type, or `bool` (which is basically an enum
+type), then it replaces the space with a set of more precise spaces that cover
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:804]: A little more like a sealed type than an enum type. A nullable type expands out to the underlying type and a `Null`-typed space. We could treat `null` like an enum value instead but I think the underlying algorithm would behave the same either way.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 4 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:816]: I don't have a strong opinion about this, but I think it's harmless to include redundant direct subtypes and I think the proposal's a little simpler and clearer to not make an exception for them.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 5 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:832]: Reworded to be from a declaration to a type.

`S` is a type, not a space (may be a confusing choice of letter). We then create a space with type `S`.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 6 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:918]: I don't think that's necessary, but I could be wrong. The proposal tries to consistently use a space of type `Null` with no constant restriction to deal with `null`.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 24]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+type* which is determined during type checking and is known for each pattern.
+The lifted space union for a pattern with matched value type `M` is:
+
+*   **Logical-or pattern:** A union of the lifted spaces of the two branches.
+
+*   **Logical-and pattern:** The intersection (defined below) of the lifted
+    space unions of the two branches.
+
+*   **Relational pattern:** The empty space union. *Relational patterns don't
+    reliably match any values, so don't help with exhaustiveness.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:365]: I suppose, yes, it does reliably match `null` when the operator is `==` or `!=`. But the current implementation doesn't seem track that for exhaustiveness, and I'm fine with it not doing so. So the proposal is consistent with that.

Today, this is not considered exhaustive:

```dart
test(int? n) => switch (n) {
  == null => 'null',
  int _ => 'int'
};
```

But this is:

```dart
test(int? n) => switch (n) {
  null => 'null',
  int _ => 'int'
};
```

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 25]
```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
...
+    List(length: 0, hasRest: false)
+    List(length: 1, hasRest: false)
+    List(length: 2, hasRest: false)
+    List(length: 3, hasRest: false)
+    List(length: 4, hasRest: false)
+    List(length: 4, hasRest: true)
     ```
 
-## Space intersection
+*   Otherwise, `s` does not expand and the result is just `s`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:962]: No, they will get unpacked incrementally as the algorithm does its depth-first traversal through the spaces.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 8 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:979]: Reworded to make it clearer that it applies to declarations.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 7 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:469]: Good question. No, we don't require constants that appear in constant patterns to have primitive equality. In practice, the only constants that actually usefully participate in exhaustiveness checking are bools and enum values, which do have primitive equality. So it mostly doesn't matter.

But it does potentially matter for reachability. Filed: https://github.com/dart-lang/sdk/issues/51873

Updated the spec here to match what that issue suggests.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 9 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:983]: I don't understand the distinction you're making here, sorry. Can you explain, or did Johhni's explanation of how overapproximation differs from I2B address this?

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 10 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:981]: You know, I don't know. I just know that the implementation does handle this.

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 26]
```diff
@@ -246,772 +267,878 @@ also be an instance of one of the known set of subtypes. That gives us the
...
     ```
-    (x: false, y: bool, z: bool) // Set x to D.
-    (x: bool, y: true, z: bool) // Set y to D.
-    (x: bool, y: bool, z: false) // Set z to D.
+    List(length: 0, hasRest: false)
+    List(length: 1, hasRest: false)
+    List(length: 2, hasRest: false)
+    List(length: 3, hasRest: false)
+    List(length: 4, hasRest: false)
+    List(length: 4, hasRest: true)
```
Comment (sgrekhov) [inline accepted/future-releases/0546-patterns/exhaustiveness.md:996]: Shouldn't it be `List(length: 5, hasRest: true)` here? `headMax` is `3` and `tailMax` is `2`, so `n==5` here

File: accepted/future-releases/0546-patterns/exhaustiveness.md [hunk 27]
```diff
@@ -185,33 +189,46 @@ we would break that principle.
...
+  Card(this.suit);
 }
 
 class Pip extends Card {
   final int pips;
+
+  Pip(this.pips, super.suit);
 }
 
 sealed class Face extends Card {}
```
[the remark this answers is not in the document]
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:217]: Nice catch. :)

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 26 above)
Comment (munificent) [reply accepted/future-releases/0546-patterns/exhaustiveness.md:996]: Oops, yes.

**Comment by munificent** [comment]
I'm going to go ahead and land this but I'm happy to iterate on it more if there are other changes to make.

File: accepted/future-releases/0546-patterns/exhaustiveness.md (hunk 8 above)
Comment (lrhn) [reply accepted/future-releases/0546-patterns/exhaustiveness.md]: I'm not sure I follow it either, but that could be because I don't actually *understand* I2B.
Or what is being attempted here.

So let me try again:

Given a *static type* (which may contain free type variables referencing surrounding declarations),
take each such occurrence of a free type variable in the type, and replace it with its default type which is:
* If the referenced type variable has a bound, the default is the overapproximation of that bound.
* Otherwise if the free type variable occurs covariantly, the default is `Object?`.
* Otherwise the default is `Never`

The resulting type has no free type variables and will be a valid super-bounded type.

... But it's not valid super-bounded. If the variable has a bound with other free type variables, the first variable occurs contravariantly, we need an *underapproximation* to use instead.

And what if it occurs invariantly, like `List<void Function<X extends T>(X)>`?


I think it can make sense to talk about type variables in static types (I'm just not used to doing so), but I think here are more details to handle, if the goal is to get a valid super-bounded type as result.
